package/MDAnalysis/analysis/waterdynamics.py

```diff
@@ -1194,99 +1194,108 @@ class SurvivalProbability(object):
 
-        if len(selected) < self.dtmax:
-            print ("ERROR: Cannot select fewer frames than dtmax")
-            return
+    def print(self, verbose, *args):
+        if self.verbose:
+            print(args)
+        elif verbose:
+            print(args)
 
-        for window_size in list(range(1, self.dtmax + 1)):
-            output = self._getMeanOnePoint(selected, window_size)
-            self.timeseries.append(output)
+    def run(self, tau_max=20, start=0, stop=-1, step=1, verbose=False):
+        """
+        Computes and returns the survival probability timeseries
+
+        Parameters
+        ----------
+        start : int
+            Zero-based index of the first frame to be analysed
+        stop : int
+            Zero-based index of the last frame to be analysed (inclusive)
+        step : int
+            Jump every `step`'th frame
+        tau_max : int
+            Survival probability is calculated for the range :math:`1 <= \tau <= tau_max`
+        verbose : Boolean
+            Overwrite the constructor's verbosity
+
+        Returns
+        -------
+        tau_timeseries : list
+            tau from 1 to tau_max
+        sp_timeseries : list
+            survival probability for each value of `tau`
+        """
 
+        # backward compatibility (and priority)
+        start = self.start if self.start is not None else start
+        stop = self.stop if self.stop is not None else stop
+        tau_max = self.tau_max if self.tau_max is not None else tau_max
 
-    def _selection_serial(self, universe, selection_str):
-        selected = []
-        pm = ProgressMeter(self.tf-self.t0, interval=10,
-                           verbose=True, offset=-self.t0)
-        for ts in universe.trajectory[self.t0:self.tf]:
-            selected.append(universe.select_atoms(selection_str))
-            pm.echo(ts.frame)
-        return selected
+        # sanity checks
+        if stop >= len(self.universe.trajectory):
+            raise ValueError("\"stop\" must be smaller than the number of frames in the trajectory.")
 
+        if stop == -1:
+            stop = self.universe.trajectory[-1].frame + 1
+        else:
+            stop = stop + 1
 
-    def _getMeanOnePoint(self, selected, window_size):
-        """
-        This function gets one point of the plot P(t) vs t. It uses the
-        _getOneDeltaPoint() function to calculate the average.
-        """
-        n = 0
-        sumDeltaP = 0.0
-        for frame_no in range(len(selected) - window_size):
-            delta = self._getOneDeltaPoint(selected, frame_no, window_size)
-            sumDeltaP += delta
-            n += 1
+        if tau_max > (stop - start):
+            raise ValueError("Too few frames selected for given tau_max.")
 
-        return sumDeltaP/n
+        # load all frames to an array of sets
+        selected_ids = []
+        for ts in self.universe.trajectory[start:stop]:
```

Comment:
add `step`?

Reply:
ah, I see you're just using step below.  So every frame is a start point, but you only do deltas of `step`?

Reply:
Hmm, not sure if I get the question. So this step here is tricky. I first load all frames (loading the right ones and considering the step makes it very tricky). 

And then during the analysis, I start the gathering the tau every step. So assume step=5 and max tau is 3. We'll have:

t = 0:
t + taus: 0, 1, 2, 3
t = 5:
t + taus: 5, 6, 7, 8, 
t= 10: .. etc

The main speed gain would be from the ensuring that frame 4, 9, ... etc would not be loaded. I will think about this how to get this right and not overcomplicate the algorithm.